server/src/main/java/org/opensearch/common/geo/GeoUtils.java

```diff
@@ -444,112 +448,201 @@ public static GeoPoint parseGeoPoint(XContentParser parser, GeoPoint point, fina
-                                geohash = subParser.text();
-                            } else {
-                                throw new OpenSearchParseException("geohash must be a string");
-                            }
-                        } else {
-                            throw new OpenSearchParseException("field must be either [{}], [{}] or [{}]", LATITUDE, LONGITUDE, GEOHASH);
-                        }
-                    } else {
-                        throw new OpenSearchParseException("token [{}] not allowed", subParser.currentToken());
+    public static GeoPoint parseGeoPoint(
+        final XContentParser parser,
+        final GeoPoint point,
+        final boolean ignoreZValue,
+        final EffectivePoint effectivePoint
+    ) throws IOException, OpenSearchParseException {
+        switch (parser.currentToken()) {
+            case START_OBJECT:
+                parseGeoPointObject(parser, point, ignoreZValue, effectivePoint);
+                break;
+            case START_ARRAY:
+                parseGeoPointArray(parser, point, ignoreZValue);
+                break;
+            case VALUE_STRING:
+                String val = parser.text();
+                point.resetFromString(val, ignoreZValue, effectivePoint);
+                break;
+            default:
+                throw new OpenSearchParseException("geo_point expected");
+        }
+        return point;
+    }
+
+    private static GeoPoint parseGeoPointObject(
+        final XContentParser parser,
+        final GeoPoint point,
+        final boolean ignoreZValue,
+        final GeoUtils.EffectivePoint effectivePoint
+    ) throws IOException {
+        try (XContentSubParser subParser = new XContentSubParser(parser)) {
+            if (subParser.nextToken() != XContentParser.Token.FIELD_NAME) {
+                throw new OpenSearchParseException(ERR_MSG_INVALID_TOKEN, subParser.currentToken());
+            }
+
+            String field = subParser.currentName();
+            if (LONGITUDE.equals(field) || LATITUDE.equals(field)) {
+                parseGeoPointObjectBasicFields(subParser, point);
+            } else if (GEOHASH.equals(field)) {
+                parseGeoHashFields(subParser, point, effectivePoint);
+            } else if (GEOJSON_TYPE.equals(field) || GEOJSON_COORDS.equals(field)) {
+                parseGeoJsonFields(subParser, point, ignoreZValue);
+            } else {
+                throw new OpenSearchParseException(ERR_MSG_INVALID_FIELDS);
+            }
+
+            if (subParser.nextToken() != XContentParser.Token.END_OBJECT) {
+                throw new OpenSearchParseException(ERR_MSG_INVALID_FIELDS);
+            }
+
+            return point;
+        }
+    }
+
+    private static GeoPoint parseGeoPointObjectBasicFields(final XContentParser parser, final GeoPoint point) throws IOException {
+        HashMap<String, Double> data = new HashMap<>();
```

Comment:
```suggestion
        Map<String, Double> data = new HashMap<>();
```

Reply:
This is a method level variable not being exposed to outside. I don't think I need to hide it here.

Reply:
IMO, still i don't think we have to expose the implementation inside method as well. I would prefer to declare interface over class irrespective of the access modifier.

Reply:
It is the actual implementation. The maps is not injected from outside or returned to outside of this method. Therefore, using Map internally does not hide anything. 

Also, if we want to use specific method of HashMap or some other Map implementation class, it end up using that specific instance type anyway.